core/src/main/scala/com/banno/kafka/consumer/ConsumerOps.scala

```diff
@@ -391,100 +418,164 @@ case class ConsumerOps[F[_], K, V](consumer: ConsumerApi[F, K, V]) {
+                   chunk.foldLeft(Map.empty[TopicPartition, Long])(_ ++ _)
+                 keepAlive.commit(offsets)
+               })
+            }
+
+          val commitStream =
+            offsetChannel.stream
+              .groupWithin(maxRecordCount0, maxElapsedTime)
+              .evalMap(doCommit)
+
+          val processStream =
+            stream
+              .evalMap(processAndEnqueueOffsets)
+              .onFinalize(
+                offsetChannel.close.void
+              ) // Signal to flush the commit stream
+
+          Stream
+            .bracket(commitStream.compile.drain.start)(_.join.void)
+            .flatMap(_ => processStream)
+        }
+    }
+  }
+
+  private sealed trait KeepAlive {
+    def commit(offsets: Map[TopicPartition, Long]): F[Unit]
+    def stream[A](gs: KeepAlive => Stream[F, A]): Stream[F, A]
+  }
+
+  private object KeepAlive {
+    def apply[A](interval: Duration)(
+        stream: KeepAlive => Stream[F, A]
+    )(implicit F: Temporal[F]): Stream[F, A] =
+      interval match {
+        case interval0: FiniteDuration =>
+          val init: F[KeepAlive] =
+            for {
+              assignments <- consumer.assignment
+              offsets <- consumer.partitionQueries.committed(
+                assignments
+              )
+              keepAlive <-
+                if (assignments.isEmpty) NoKeepAlive.pure[F]
+                else
+                  for {
+                    lock <- Semaphore(1)
+                    ref <- Ref.of(
+                      offsets.view.mapValues(_.offset).toMap
+                    )
+                  } yield HasKeepAlive(interval0, lock, ref)
+            } yield keepAlive
+
+          Stream.eval(init).flatMap(_.stream(stream))
+        case _ => stream(NoKeepAlive)
       }
 
-  case class OffsetCommitState(
-      offsets: Map[TopicPartition, Long],
-      recordCount: Long,
-      lastCommitTime: FiniteDuration,
-  ) {
-    def update(os: Map[TopicPartition, Long], count: Long): OffsetCommitState =
-      copy(
-        offsets = offsets ++ os,
-        recordCount = recordCount + count,
-      )
+    case object NoKeepAlive extends KeepAlive {
+      def commit(offsets: Map[TopicPartition, Long]): F[Unit] = {
+        val nextOffsets =
+          offsets.view.mapValues(o => new OffsetAndMetadata(o + 1)).toMap
 
-    def needToCommit(
-        maxRecordCount: Long,
-        now: FiniteDuration,
-        maxElapsedTime: FiniteDuration,
-    ): Option[Map[TopicPartition, OffsetAndMetadata]] = {
-      (recordCount >= maxRecordCount || (now - lastCommitTime) >= maxElapsedTime)
-        .guard[Option]
-        .as(nextOffsets)
+        consumer.commitSync(nextOffsets)
+      }
+
+      def stream[A](gs: KeepAlive => Stream[F, A]): Stream[F, A] = gs(this)
     }
 
-    def nextOffsets: Map[TopicPartition, OffsetAndMetadata] =
-      offsets.view.mapValues(o => new OffsetAndMetadata(o + 1)).toMap
+    case class HasKeepAlive(
```

Comment:
The `KeepAlive` trait is private, so should `HasKeepAlive` also be explicitly made private?

Reply:
See above :smile: